File: inferlet-js/src/tokenizer.ts
// Tokenizer class for text tokenization and detokenization
// Mirrors the Rust Tokenizer from inferlet/src/lib.rs

import { Tokenizer as TokenizerBinding, getTokenizer } from 'inferlet:core/tokenize';
import type { Model } from 'inferlet:core/common';

/**
 * Tokenizer class that wraps the WIT tokenize bindings
 * Provides methods to convert text to token IDs and vice versa
 */
export class Tokenizer {
  private readonly inner: TokenizerBinding;

  /**
   * Creates a new Tokenizer instance for the given model
   * @param model - The Model instance to create a tokenizer for
   */
  constructor(model: Model) {
    this.inner = getTokenizer(model);
  }

  /**
   * Converts a string of text into a sequence of token IDs.
   *
   * @param text - The input string to tokenize
   * @returns A Uint32Array containing the corresponding token IDs
   */
  tokenize(text: string): Uint32Array {
    return this.inner.tokenize(text);
  }

  /**
   * Converts a sequence of token IDs back into a human-readable string.
   *
   * @param tokens - A Uint32Array of token IDs to detokenize
   * @returns The reconstructed string
   */
  detokenize(tokens: Uint32Array): string {
    return this.inner.detokenize(tokens);
  }

  /**
   * The entire vocabulary of the tokenizer.
   * Returns a tuple of [token IDs, byte sequences]
   */
  get vocabs(): [Uint32Array, Array<Uint8Array>] {
    return this.inner.getVocabs();
  }

  /**
   * The special tokens recognized by the tokenizer.
   * Returns a tuple of [special token IDs, byte sequences]
   */
  get specialTokens(): [Uint32Array, Array<Uint8Array>] {
    return this.inner.getSpecialTokens();
  }

  /**
   * The split regular expression used by the tokenizer.
   */
  get splitRegex(): string {
    return this.inner.getSplitRegex();
  }

  // Deprecated methods for backward compatibility
  /** @deprecated Use `vocabs` getter instead */
  getVocabs(): [Uint32Array, Array<Uint8Array>] { return this.vocabs; }
  /** @deprecated Use `specialTokens` getter instead */
  getSpecialTokens(): [Uint32Array, Array<Uint8Array>] { return this.specialTokens; }
  /** @deprecated Use `splitRegex` getter instead */
  getSplitRegex(): string { return this.splitRegex; }
}

File: sdk/python/src/inferlet_py/tokenizer.py
"""
Tokenizer class for encoding and decoding text.
"""

from wit_world.imports import tokenize as _tokenize


class Tokenizer:
    """
    Tokenizer for encoding text to tokens and decoding tokens to text.
    """

    def __init__(self, inner: _tokenize.Tokenizer) -> None:
        self._inner = inner

    def encode(self, text: str) -> list[int]:
        """
        Encode text to a list of token IDs.

        Args:
            text: The text to encode

        Returns:
            List of token IDs
        """
        return list(self._inner.tokenize(text))

    def decode(self, tokens: list[int]) -> str:
        """
        Decode token IDs back to text.

        Args:
            tokens: List of token IDs

        Returns:
            Decoded text string
        """
        return self._inner.detokenize(tokens)

    def __call__(self, text: str) -> list[int]:
        """
        Shorthand for encode.

        Example:
            tokens = tokenizer('hello')  # -> [1, 2, 3]
        """
        return self.encode(text)

    @property
    def vocabs(self) -> tuple[list[int], list[bytes]]:
        """
        Get the tokenizer's vocabulary.

        Returns:
            Tuple of (token IDs, token byte sequences)
        """
        ids, tokens = self._inner.get_vocabs()
        return list(ids), list(tokens)

    @property
    def split_regex(self) -> str:
        """Get the split regular expression used by the tokenizer."""
        return self._inner.get_split_regex()

    @property
    def special_tokens(self) -> tuple[list[int], list[bytes]]:
        """
        Get the special tokens recognized by the tokenizer.

        Returns:
            Tuple of (token IDs, token byte sequences)
        """
        ids, tokens = self._inner.get_special_tokens()
        return list(ids), list(tokens)
